db/migrate/20190108031411_add_image_name_to_operating_systems.rb

```diff
@@ -0,0 +1,157 @@
+    belongs_to :vm_or_template
+    belongs_to :computer_system
+
+    # Using a `before_save` here since this is the mechanism that will be used
+    # in the app.  Causes a bit of issues in the specs, but proves that this
+    # would work moving forward.
+    before_save :update_platform_and_image_name
+
+    def update_platform_and_image_name
+      obj = case
+            when host_id            then host
+            when vm_or_template_id  then vm_or_template
+            when computer_system_id then computer_system
+            else
+              StubOperatingSystemObj.new(self)
+            end
+
+      if obj
+        self.image_name = self.class.image_name(obj)
+        self.platform   = self.image_name.split("_").first
+      end
+    end
+
+    @@os_map = [ # rubocop:disable Style/ClassVars
+      ["windows_generic", %w(winnetenterprise w2k3 win2k3 server2003 winnetstandard servernt)],
+      ["windows_generic", %w(winxppro winxp)],
+      ["windows_generic", %w(vista longhorn)],
+      ["windows_generic", %w(win2k win2000)],
+      ["windows_generic", %w(microsoft windows winnt)],
+      ["linux_ubuntu",    %w(ubuntu)],
+      ["linux_chrome",    %w(chromeos)],
+      ["linux_chromium",  %w(chromiumos)],
+      ["linux_suse",      %w(suse sles)],
+      ["linux_redhat",    %w(redhat rhel)],
+      ["linux_fedora",    %w(fedora)],
+      ["linux_gentoo",    %w(gentoo)],
+      ["linux_centos",    %w(centos)],
+      ["linux_debian",    %w(debian)],
+      ["linux_coreos",    %w(coreos)],
+      ["linux_esx",       %w(vmnixx86 vmwareesxserver esxserver vmwareesxi)],
+      ["linux_solaris",   %w(solaris)],
+      ["linux_generic",   %w(linux)]
+    ]
+
+    # rubocop:disable Naming/VariableName
+    def self.normalize_os_name(osName)
+      findStr = osName.downcase.gsub(/[^a-z0-9]/, "")
+      @@os_map.each do |a|
+        a[1].each do |n|
+          return a[0] unless findStr.index(n).nil?
+        end
+      end
+      "unknown"
+    end
+    # rubocop:enable Naming/VariableName
+
+    # rubocop:disable Naming/VariableName
+    def self.image_name(obj)
+      osName = nil
+
+      # Select most accurate name field
+      os = obj.operating_system
+      if os
+        # check the given field names for possible matching value
+        osName = [:distribution, :product_type, :product_name].each do |field| # rubocop:disable Style/SymbolArray
+          os_field = os.send(field)
+          break(os_field) if os_field && OperatingSystem.normalize_os_name(os_field) != "unknown"
+        end
+
+        # If the normalized name comes back as unknown, nil out the value so we can get it from another field
+        if osName.kind_of?(String)
+          osName = nil if OperatingSystem.normalize_os_name(osName) == "unknown"
+        else
+          osName = nil
+        end
+      end
+
+      # If the OS Name is still blank check the 'user_assigned_os'
+      if osName.nil? && obj.respond_to?(:user_assigned_os) && obj.user_assigned_os
```

Comment:
This is probably a naive suggestion but...

```suggestion
      osName ||= obj.try(:user_assigned_os)
```

Reply:
Still same...

https://github.com/ManageIQ/manageiq/blob/184bb9d3864d84b91903bc0003a9d77362f5863c/app/models/operating_system.rb#L83